langchain/callbacks/mlflow_callback.py

```diff
@@ -0,0 +1,691 @@
+        on_llm_end_records_df = pd.DataFrame(self.records["on_llm_end_records"])
+
+        llm_input_prompts_df = (
+            on_llm_start_records_df[["step", "prompt", "name"]]
+            .dropna(axis=1)
+            .rename({"step": "prompt_step"}, axis=1)
+        )
+        complexity_metrics_columns = []
+        visualizations_columns = []
+
+        complexity_metrics_columns = [
+            "flesch_reading_ease",
+            "flesch_kincaid_grade",
+            "smog_index",
+            "coleman_liau_index",
+            "automated_readability_index",
+            "dale_chall_readability_score",
+            "difficult_words",
+            "linsear_write_formula",
+            "gunning_fog",
+            "text_standard",
+            "fernandez_huerta",
+            "szigriszt_pazos",
+            "gutierrez_polini",
+            "crawford",
+            "gulpease_index",
+            "osman",
+        ]
+
+        visualizations_columns = ["dependency_tree", "entities"]
+
+        llm_outputs_df = (
+            on_llm_end_records_df[
+                [
+                    "step",
+                    "text",
+                    "token_usage_total_tokens",
+                    "token_usage_prompt_tokens",
+                    "token_usage_completion_tokens",
+                ]
+                + complexity_metrics_columns
+                + visualizations_columns
+            ]
+            .dropna(axis=1)
+            .rename({"step": "output_step", "text": "output"}, axis=1)
+        )
+        session_analysis_df = pd.concat([llm_input_prompts_df, llm_outputs_df], axis=1)
+        session_analysis_df["chat_html"] = session_analysis_df[
+            ["prompt", "output"]
+        ].apply(
+            lambda row: construct_html_from_prompt_and_generation(
+                row["prompt"], row["output"]
+            ),
+            axis=1,
+        )
+        return session_analysis_df
+
+    def flush_tracker(self, langchain_asset: Any = None, finish: bool = False):
+
+        pd = import_pandas()
+        self.mlflg.table("action_records", pd.DataFrame(self.records["action_records"]))
+        session_analysis_df = self._create_session_analysis_df()
+        chat_html = session_analysis_df.pop("chat_html")
+        chat_html = chat_html.replace("\n", "", regex=True)
+        self.mlflg.table("session_analysis", pd.DataFrame(session_analysis_df))
+        self.mlflg.html("".join(chat_html.tolist()), "chat_html")
+
+        if langchain_asset:
+            langchain_asset_path = Path(self.temp_dir.name, "model.json")
+            try:
+                langchain_asset.save(langchain_asset_path)
+            except ValueError:
+                langchain_asset.save_agent(langchain_asset_path)
+            except NotImplementedError as e:
+                print("Could not save model.")
+                print(repr(e))
+                pass
+            self.mlflg.artifact(langchain_asset_path)

```

Comment:
I'm not sure if you're planning on coordinating merging this PR with @liangz1 after the flavor is implemented (seems like that would be the easiest), but if not you can implement a flavor in "userspace" something like this: https://github.com/akshaya-a/langchain/pull/1/files#diff-a8185aaebbb28c2ebd37436cf39390e57ede177bc301cb7dc897b90f11d3b2bbR134

Reply:
Hi @akshaya-a , I tried out your feature branch, it works well with some minor fixes and the logged data looks awesome. Would you like to file a PR for it? Besides, it would be great if you can use the langchain flavor released with mlflow 2.3 to log the model.

Reply:
hey @liangz1 , no I think @mahmedk's PR is similar enough and he's responsive with feedback so I'm planning on just abandoning my PR. Hope that's ok with you both! Also FYI I'm AK on the AzureML team working with Jin + your team so we're in sync :)

@mahmedk let me know if you want me to contribute any specific things to your PR (the maintainer seems to be mainly awaiting the sample notebook like wandb) to parallelize efforts.

Reply:
@akshaya-a thanks ! Yes, am pending to complete the notebook. Haven't been able to spend time on it. If you could do that or if you have already done it, please point me to it, I will use it and test the PR.